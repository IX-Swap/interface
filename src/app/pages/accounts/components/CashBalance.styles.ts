import makeStyles from '@mui/styles/makeStyles'

export const useStyles = makeStyles(theme => {
  return {
    wrapper: {
      borderRadius: 8,
      padding: theme.spacing(3),
      boxSizing: 'border-box',

      '& .MuiFilledInput-root': {
        width: 'max-content',
        height: 'max-content!important',
        paddingRight: theme.spacing(1.5625),
        color: theme.palette.text.secondary,

        '&:hover': {
          backgroundColor: 'initial'
        },

        '&::before': {
          display: 'none'
        },

        '&::after': {
          display: 'none'
        }
      }
    },
    valueBlock: {
      fontSize: 24,
      fontWeight: 600,
      display: 'block',
      margin: '0 auto',
      textAlign: 'center'
    },
    valueNoSelect: {
      margin: '8px auto 0'
    },
    currency: {
      fontSize: 'inherit',
      fontWeight: 'inherit',
      display: 'inline',
      color: theme.palette.text.secondary
    },
    withoutBackground: {
      padding: 0,
      backgroundColor: 'none'
    },
    icon: {
      top: 12.5
    },
    iconOpen: {
      fill: '#4C88FF!important'
    },
    filled: {
      paddingTop: theme.spacing(1.25),
      paddingLeft: 0,
      '&:focus': {
        backgroundColor: 'initial'
      }
    },
    menuPaper: {
      minWidth: '120px!important',
      maxWidth: 120,
      marginLeft: theme.spacing(6.25),
      paddingTop: '8px!important',
      paddingBottom: '8px!important',
      marginTop: '0!important'
    },
    menuList: {
      padding: 0,

      '& .MuiMenuItem-root': {
        paddingTop: theme.spacing(1.5),
        paddingBottom: theme.spacing(1.5),
        '&:focus': {
          backgroundColor: 'initial'
        }
      }
    },
    button: {
      minWidth: 130,
      height: 40,

      [theme.breakpoints.up('sm')]: {
        height: 50
      }
    }
  }
})
