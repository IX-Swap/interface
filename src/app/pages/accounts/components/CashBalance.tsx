import { Grid, Paper, Typography } from '@mui/material'
import { useVirtualAccount } from 'app/pages/accounts/hooks/useVirtualAccount'
import { renderMoney } from 'helpers/numbers'
import React, { useState } from 'react'
import { Icon } from 'ui/Icons/Icon'
import { InputLabel } from 'ui/Select/InputLabel/InputLabel'
import { Select } from 'ui/Select/Select'
import { SelectItem } from 'ui/Select/SelectItem/SelectItem'
import { useStyles } from './CashBalance.styles'

export const CashBalance = () => {
  const classes = useStyles()
  const [virtualAccountId, setVirtualAccountId] = useState<string | undefined>(
    undefined
  )
  const { data, list, isLoading } = useVirtualAccount(virtualAccountId)

  if (isLoading || data === undefined) {
    return null
  }

  const getSumBalanceOfAllAccounts = () => {
    const sumOfSGDValues = list.reduce(
      (acc, item) => acc + (item.balance.sgdValue ?? 0),
      0
    )
    const sumOfUSDValues = list.reduce(
      (acc, item) => acc + (item.balance.usdValue ?? 0),
      0
    )

    return data.currency === 'SGD' ? sumOfSGDValues : sumOfUSDValues
  }

  return (
    <Paper className={classes.wrapper}>
      <Grid container alignItems={'center'} justifyContent={'center'}>
        <Grid item flexBasis={'auto'}>
          <InputLabel style={{ marginBottom: 0 }}>
            <Typography color={'text.secondary'} textAlign={'center'}>
              Available Balance
            </Typography>
          </InputLabel>
          <Select
            variant='filled'
            defaultValue={data.currency}
            classes={{
              icon: classes.icon,
              iconOpen: classes.iconOpen,
              filled: classes.filled
            }}
            MenuProps={{
              classes: { paper: classes.menuPaper, list: classes.menuList }
            }}
            renderValue={() => (
              <Typography fontSize={24} fontWeight={600} display={'inline'}>
                {renderMoney(getSumBalanceOfAllAccounts(), '')}{' '}
                <Typography
                  fontSize={'inherit'}
                  fontWeight={'inherit'}
                  display={'inline'}
                  color={'text.secondary'}
                >
                  {data.currency}
                </Typography>
              </Typography>
            )}
            IconComponent={props => <Icon name='chevron-down' {...props} />}
          >
            {list.map(item => (
              <SelectItem
                key={item._id}
                value={item.currency}
                onClick={() => setVirtualAccountId(item.accountNumber)}
              >
                {item.currency}
              </SelectItem>
            ))}
          </Select>
        </Grid>
      </Grid>
    </Paper>
  )
}
