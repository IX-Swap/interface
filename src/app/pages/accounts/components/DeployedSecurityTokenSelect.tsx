import { SelectProps, FormControl } from '@mui/material'
import { useAssetsData } from 'hooks/asset/useAssetsData'
import React from 'react'
import { InputLabel } from 'ui/Select/InputLabel/InputLabel'
import { Autocomplete } from 'ui/Select/Autocomplete/Autocomplete'
import { SecurityTokenSelectItem } from 'ui/Select/SelectItem/SecurityToken/SecurityToken'

export const DeployedSecurityTokenSelect = React.forwardRef(
  (props: SelectProps) => {
    const { data, isLoading } = useAssetsData(
      props.type ?? 'Security',
      500,
      true
    )

    const hasTokens = !isLoading && data !== undefined && data.list.length > 0
    const options = !hasTokens
      ? []
      : data?.list?.map(token => {
          return {
            label: [token?.symbol, token?.name, token?.network?.name],
            render: <SecurityTokenSelectItem sto={token} />,
            value: token
          }
        })

    return (
      <FormControl fullWidth variant='outlined'>
        <InputLabel>{props.label ?? 'Select Security Token'}</InputLabel>

        <Autocomplete
          {...props}
          placeholder={
            hasTokens
              ? // eslint-disable-next-line @typescript-eslint/restrict-template-expressions, @typescript-eslint/no-base-to-string
                `Select ${props.label ?? 'Select Security Token'}`
              : 'No tokens available'
          }
          disabled={isLoading || !hasTokens}
          options={options}
        />
      </FormControl>
    )
  }
)

DeployedSecurityTokenSelect.defaultProps = {
  label: 'No tokens'
}

DeployedSecurityTokenSelect.displayName = 'Select_DeployedSecurityTokenSelect'
