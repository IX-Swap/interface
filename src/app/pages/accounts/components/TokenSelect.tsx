import { MenuItem, Select, SelectProps } from '@mui/material'
import { useGetCustody } from 'app/pages/accounts/hooks/useGetCustody'
import React from 'react'
import { AssetBalance } from 'types/balance'
import { isEmptyString } from 'helpers/strings'

export const TokenSelect = React.forwardRef((props: SelectProps, ref) => {
  const { data, isLoading } = useGetCustody()
  const label = isLoading || data === undefined ? 'No tokens' : props.label

  return (
    <Select {...props} label={label} ref={ref}>
      <MenuItem disabled value={undefined}>
        {label}
      </MenuItem>
      {data?.map((token: AssetBalance) => (
        <MenuItem key={token.symbol} value={token.symbol}>
          {token.symbol} {!isEmptyString(token.name) && `(${token.name})`}
        </MenuItem>
      ))}
    </Select>
  )
})

TokenSelect.defaultProps = {
  label: 'No tokens'
}
