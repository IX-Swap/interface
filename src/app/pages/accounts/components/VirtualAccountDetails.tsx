import { Grid } from '@material-ui/core'
import { VirtualAccountCard } from 'app/pages/accounts/components/VirtualAccountCard/VirtualAccountCard'
import { VirtualAccountNumberInfo } from 'app/pages/accounts/components/VirtualAccountCard/VirtualAccountNumberInfo'
import { useVirtualAccount } from 'app/pages/accounts/hooks/useVirtualAccount'
import { AutoAssignVirtualAccountForm } from 'app/pages/accounts/pages/banks/components/AutoAssignVirtualAccountForm/AutoAssignVirtualAccountForm'
import React from 'react'

export interface VirtualAccountDetailsProps {
  virtualAccountId?: string
  showAddForm?: boolean
}

export const VirtualAccountDetails = ({
  virtualAccountId,
  showAddForm = false
}: VirtualAccountDetailsProps) => {
  const { data, list, isLoading } = useVirtualAccount(virtualAccountId)

  if (isLoading || data === undefined) {
    return null
  }

  return (
    <Grid container spacing={3}>
      <Grid item>
        <VirtualAccountCard
          label={data.user.name}
          info={
            <VirtualAccountNumberInfo
              accountNumber={data.accountNumber}
              currency={data.currency}
              availableBalance={data.balance.available}
              onHold={data.balance.onHold}
            />
          }
        />
      </Grid>
      {showAddForm ? (
        <Grid item>
          <AutoAssignVirtualAccountForm isAdditional={list?.length === 1} />
        </Grid>
      ) : null}
    </Grid>
  )
}
