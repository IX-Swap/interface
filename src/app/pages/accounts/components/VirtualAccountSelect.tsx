import {
  FormControlLabel,
  Grid,
  Radio,
  RadioGroup,
  RadioGroupProps,
  Typography
} from '@mui/material'
import { useVirtualAccount } from 'app/pages/accounts/hooks/useVirtualAccount'
import React from 'react'
import { VirtualAccount } from 'types/virtualAccount'

export interface VirtualAccountSelectProps extends RadioGroupProps {
  customLabel?: string
}

export const VirtualAccountSelect = (props: VirtualAccountSelectProps) => {
  const { list } = useVirtualAccount()

  if (list === undefined || list.length < 1) {
    return null
  }

  return (
    <RadioGroup
      defaultValue={props.defaultValue}
      onChange={props.onChange}
      value={props.value}
    >
      <Grid container spacing={2} alignContent='center'>
        <Grid item xs={12}>
          <Typography variant='subtitle1'>{props.customLabel}</Typography>
        </Grid>
        {list.map((item: VirtualAccount) => (
          <Grid item key={item._id}>
            <FormControlLabel
              label={`${item.accountNumber} (${item.currency})`}
              value={item.accountNumber}
              control={<Radio />}
            />
          </Grid>
        ))}
      </Grid>
    </RadioGroup>
  )
}
