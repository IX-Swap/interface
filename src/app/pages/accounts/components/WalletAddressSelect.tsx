import { SelectProps } from '@mui/material'
import { useWalletAddresses } from 'app/pages/accounts/hooks/useWalletAddresses'
import React from 'react'
import { WithdrawalAddress } from 'types/withdrawalAddress'
import { InputLabel } from 'ui/Select/InputLabel/InputLabel'
import { Select } from 'ui/Select/Select'
import { SelectItem } from 'ui/Select/SelectItem/SelectItem'

export const WalletAddressSelect = React.forwardRef(
  (props: SelectProps, ref) => {
    const { data, isLoading } = useWalletAddresses()

    if (data === undefined || isLoading) {
      return null
    }

    return (
      <>
        <InputLabel>{props.label}</InputLabel>
        <Select
          {...props}
          ref={ref}
          label={undefined}
          placeholder={String(props.label)}
          displayEmpty
        >
          <SelectItem disabled value={undefined}>
            {props.label}
          </SelectItem>
          {data.map((wallet: WithdrawalAddress) => (
            <SelectItem key={wallet.address} value={wallet.address}>
              {wallet.label}
            </SelectItem>
          ))}
        </Select>
      </>
    )
  }
)
WalletAddressSelect.displayName = 'Select_ WalletAddressSelect'
