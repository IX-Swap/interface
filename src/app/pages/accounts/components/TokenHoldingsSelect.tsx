import { SelectProps, FormControl, Box, Button } from '@mui/material'
// import { useGetCustody } from 'app/pages/accounts/hooks/useGetCustody'
import { useGetTokenHoldings } from 'app/pages/accounts/hooks/useGetTokenHoldings'
import React from 'react'
import { InputLabel } from 'ui/Select/InputLabel/InputLabel'
import { Autocomplete } from 'ui/Select/Autocomplete/Autocomplete'
import { SecurityTokenSelectItem } from 'ui/Select/SelectItem/SecurityToken/SecurityToken'
import { AppRouterLinkComponent } from 'components/AppRouterLink'
import { InvestRoute } from 'app/pages/invest/router/config'

export const TokenHoldingsSelect = React.forwardRef((props: SelectProps) => {
  const { data, isLoading } = useGetTokenHoldings(props.type ?? 'Security')

  const hasTokens = !isLoading && data !== undefined && data.length > 0
  const options = !hasTokens
    ? []
    : data?.map(token => {
        return {
          label: [
            token?.asset?.symbol,
            token?.asset?.name,
            token?.asset?.network?.name
          ],
          render: <SecurityTokenSelectItem sto={token?.asset} />,
          value: token
        }
      })

  return (
    <FormControl fullWidth variant='outlined'>
      <InputLabel>{props.label ?? 'Select Security Token'}</InputLabel>

      <Box display={'flex'} gap={2}>
        <Autocomplete
          {...props}
          placeholder={
            hasTokens
              ? // eslint-disable-next-line @typescript-eslint/restrict-template-expressions, @typescript-eslint/no-base-to-string
                `Select ${props.label ?? 'Select Security Token'}`
              : `You do not have a ${
                  props.type === 'Security' ? 'security token' : 'stablecoin'
                }`
          }
          disabled={isLoading || !hasTokens}
          options={options}
        />

        {!hasTokens && (
          <Button
            variant='contained'
            size='medium'
            sx={{ whiteSpace: 'nowrap' }}
            disableElevation
            component={AppRouterLinkComponent}
            to={InvestRoute.primaryOfferings}
          >
            Invest in STO
          </Button>
        )}
      </Box>
    </FormControl>
  )
})

TokenHoldingsSelect.defaultProps = {
  label: 'No tokens'
}

TokenHoldingsSelect.displayName = 'Select_TokenHoldingsSelect'
