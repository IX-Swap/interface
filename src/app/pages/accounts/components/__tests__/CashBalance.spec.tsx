import * as useVirtualAccount from 'app/pages/accounts/hooks/useVirtualAccount'
import React from 'react'
import { render } from 'test-utils'
import { generateQueryResult } from '__fixtures__/useQuery'
import { virtualAccountsSample } from '__fixtures__/virtualAccounts'
import { CashBalance } from 'app/pages/accounts/components/CashBalance'

describe('CashBalance', () => {
  afterEach(async () => {
    jest.clearAllMocks()
  })

  it('returns null when isLoading is true', () => {
    const useVirtualAccountsResponse = generateQueryResult({
      data: virtualAccountsSample[0],
      isLoading: true
    })

    jest
      .spyOn(useVirtualAccount, 'useVirtualAccount')
      .mockImplementation(() => useVirtualAccountsResponse as any)
    const { container } = render(<CashBalance />)

    expect(container).toBeEmptyDOMElement()
  })
  it('returns null when data is undefined', () => {
    const useVirtualAccountsResponse = generateQueryResult({
      data: undefined,
      isLoading: false
    })

    jest
      .spyOn(useVirtualAccount, 'useVirtualAccount')
      .mockImplementation(() => useVirtualAccountsResponse as any)
    const { container } = render(<CashBalance />)

    expect(container).toBeEmptyDOMElement()
  })

  it('should match snapshot', () => {
    const useVirtualAccountsResponse = generateQueryResult({
      data: virtualAccountsSample[0],

      isLoading: false
    } as any)

    jest.spyOn(useVirtualAccount, 'useVirtualAccount').mockImplementation(
      () =>
        ({
          ...useVirtualAccountsResponse,
          list: virtualAccountsSample
        } as any)
    )
    const { container } = render(<CashBalance />)

    expect(container).toMatchSnapshot()
  })
})
