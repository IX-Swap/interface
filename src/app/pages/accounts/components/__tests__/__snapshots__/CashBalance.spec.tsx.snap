// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CashBalance should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 makeStyles-wrapper-17 css-11gelwl-MuiPaper-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-1402ce4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-auto css-y4b3lo-MuiGrid-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-root MuiFormLabel-colorPrimary makeStyles-wrapper-25 css-1oqcrp7-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          style="margin-bottom: 0px;"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-z3lpfj-MuiTypography-root"
          >
            Available Balance
          </p>
        </label>
        <div
          class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth makeStyles-wrapper-26 css-1k0iil1-MuiInputBase-root-MuiFilledInput-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            class="makeStyles-select-27 MuiSelect-select makeStyles-filled-21 MuiSelect-filled MuiFilledInput-input MuiInputBase-input css-a96h1p-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
            role="button"
            tabindex="0"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-elf1yz-MuiTypography-root"
            >
               246
               
              <p
                class="MuiTypography-root MuiTypography-body1 css-yh93o2-MuiTypography-root"
              >
                USD
              </p>
            </p>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="USD"
          />
          <svg
            class="makeStyles-icon-19 MuiSelect-icon MuiSelect-iconFilled css-jukrun-MuiSelect-icon"
            fill="#778194"
            height="24px"
            width="24px"
          >
            <use
              href="image-mock#chevron-down"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
